evm/vm/forks/sharding/__init__.py

```diff
@@ -0,0 +1,266 @@
+from evm.constants import (
+    ENTRY_POINT,
+    STACK_DEPTH_LIMIT,
+)
+
+from evm.exceptions import (
+    InsufficientFunds,
+    StackDepthLimit,
+    ContractCreationCollision,
+    OutOfGas,
+)
+
+from evm.vm.message import (
+    ShardingMessage,
+)
+from evm.vm.computation import (
+    Computation,
+)
+
+from evm.utils.address import (
+    generate_create2_contract_address,
+)
+from evm.utils.hexadecimal import (
+    encode_hex,
+)
+from evm.utils.keccak import (
+    keccak,
+)
+
+from ..byzantium import ByzantiumVM
+from ..spurious_dragon.constants import (
+    EIP170_CODE_SIZE_LIMIT,
+    GAS_CODEDEPOSIT,
+)
+from ..frontier.constants import (
+    REFUND_SELFDESTRUCT,
+)
+from .validation import validate_sharding_transaction
+from .blocks import ShardingBlock
+
+
+def _execute_sharding_transaction(vm, transaction):
+    #
+    # 1) Pre Computation
+    #
+
+    # Validate the transaction
+    transaction.validate()
+
+    vm.validate_transaction(transaction)
+
+    gas_fee = transaction.gas * transaction.gas_price
+    with vm.state_db() as state_db:
+        # Buy Gas
+        state_db.delta_balance(transaction.to, -1 * gas_fee)
+
+        # Setup VM Message
+        message_gas = transaction.gas - transaction.intrensic_gas
+
+        if transaction.code:
+            contract_address = generate_create2_contract_address(
+                b'',
+                transaction.code,
+            )
+            data = b''
+            code = transaction.code
+        else:
+            contract_address = None
+            data = transaction.data
+            code = state_db.get_code(transaction.to)
+
+    vm.logger.info(
+        (
+            "TRANSACTION: to: %s | gas: %s | "
+            "gas-price: %s | data-hash: %s | code-hash: %s"
+        ),
+        encode_hex(transaction.to),
+        transaction.gas,
+        transaction.gas_price,
+        encode_hex(keccak(transaction.data)),
+        encode_hex(keccak(transaction.code)),
+    )
+
+    message = ShardingMessage(
+        gas=message_gas,
+        gas_price=transaction.gas_price,
+        to=transaction.to,
+        sender=ENTRY_POINT,
+        value=0,
+        data=data,
+        code=code,
+        create_address=contract_address,
+    )
+
+    #
+    # 2) Apply the message to the VM.
+    #
+    if message.is_create:
```

Comment:
Right now `message.is_create` is `True` even if `transaction.code` is empty (as in this case `message.code` is set to the contract code in line 70). So I think there's something wrong here, but I'm not sure what's the best way to fix it.

Reply:
ah that's also true, it's my mistake. And I don't see a clear and simple way around this either. Perhaps have `message.is_create` set by `execute_transaction` and `CREATE*` opcodes?

Reply:
Yeah, that sounds reasonable.

Reply:
I'm not sure if I'm reading this right, but if what you're suggesting involves mutating the `Message` object I would like to find another way as I think `Message` should be treated as immutable.

Reply:
That makes sense. How about setting the `message.is_create` flag during message initialization? Since we can already pass in `create_address`, we can 
1. simply add a new `is_create` parameter 
OR
2. tell by checking the value of `create_address`
OR
3. just replace `create_address` with `is_create` (since in Account Abstraction, instead of `CREATE_CONTRACT_ADDRESS`, `to` should be set as the address of the new contract. Setting `storage_address` to `create_address` would be the same as leave it empty).

Reply:
I'd say go for 2 or 3; keep it simple.